### PR TITLE
add intermediate solutions

### PR DESCRIPTION
**Pull request by nobod4t**
@laike9m Please help review. Thanks.

#19

### CODE REVIEW

**Comment by laike9m** [comment]
Thanks a lot, taking a look now. Assuming you've tested them yourself (including the other PR)?

File: challenges/intermediate-decorator/solution.py
Comment (laike9m) [inline challenges/intermediate-decorator/solution.py:1]: This one seems incorrect, [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDRQAqi1AakSAFCiRTACuKAMYwwYEgGdM2XPgDuIIgnEcODKAF5GzNiAAUAIgb76FMAIAm64mUo0AlCvPVgUJ4NxERe-kIBcjOygAWgA%2BRl8VKCioAAF5RXFdH0EHaNdnKHiEBGo9ACp2NHF6PLyAa1lC8TsItLSQahg%2BEBReAUFdApAikvLK7urI6IamlsyFbNyVAGJpqABRFHMoMBc4MxAodycoWY4Ytw8vDicXYFFdIn9UGBKTf3EYEEDQqAA5MBRqWuiAOn%2BVPtDgpjqcoBR2LpBNcULdXA8ni8wh8vj8ov9foDzmBdABGEzqfQAJn0DghekE6nxrkJJIcXAu1OJpKic2oAA8csIgvAckFcuBONjLlT6NRaSzdlAOVyYDzmPyQIKTtR3CDcHjAlKZaq5bzqIrBUA) reports

> Expression of type "_Wrapped[..., Unknown, (*args: Unknown, **kwargs: Unknown), Unknown]" cannot be assigned to return type "T@decorator"
  Type "_Wrapped[..., Unknown, (*args: Unknown, **kwargs: Unknown), Unknown]" cannot be assigned to type "T@decorator" 

Replacing `return wrapper` with `return cast(T, wrapper)` make tests passed, but I'm not sure if that's the optimal solution. @frostming what did you have in mind when creating this one?

File: challenges/intermediate-generic-with-constraints/solution.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+"""
+TODO:
+
+Define a function that accepts two parameters of the same type and returns the same type.
+The type can only be str or int.
+"""
+from typing import TypeVar
+
+T = TypeVar("T", int, str)
```
Comment (laike9m) [inline challenges/intermediate-generic-with-constraints/solution.py:9]: Let's add an alternative solution for 3.12 and above
```
def add[T: (str,int)](a: T, b: T) -> T:
    return a
```

File: challenges/intermediate-union/solution.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+"""
+TODO:
+    foo should accept a argument that's either a string or integer.
+"""
+from typing import Union
+
+
+def foo(x: Union[str, int]) -> None:
+    pass
+
```
Comment (laike9m) [inline challenges/intermediate-union/solution.py:10]: let's also add `str | int`

File: challenges/intermediate-generic/solution.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+The function `add` accepts two parameters of the same type and returns the same type.
+"""
+
+from typing import TypeVar
+
+T = TypeVar('T')
+def add(a, b: T) -> T:
+    return a
+
```
Comment (laike9m) [inline challenges/intermediate-generic/solution.py:12]: ditto, add this too
```
def add[T](a, b: T) -> T:
    return a
```

File: challenges/intermediate-decorator/solution.py
Comment (frostming) [reply challenges/intermediate-decorator/solution.py:1]: Oh no, the question without any change already passes the test.
```python
def decorator(func: T) -> T:
    return func
```

Keep in mind that this challenge only examines the correctness of types and does not focus on the correctness of implementation.@nobod4t can you do the following changes?

1. Remove the type info from the question code for this challenge
2. Make sure there is no other challenges as such, namely, the question code shouldn't pass. Ideally, there shouldn't be any type annotation in the function or object to be tested.

File: challenges/intermediate-generic/solution.py (hunk 3 above)
Comment (frostming) [reply challenges/intermediate-generic/solution.py:12]: By design `a` should also be `T`, if it passes the test then the test doesn't cover.

Comment (laike9m) [reply challenges/intermediate-generic/solution.py:12]: Sg, we can add the test case first

File: challenges/intermediate-generic-with-constraints/solution.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+def add(a, b: T) -> T:
```
Comment (frostming) [inline challenges/intermediate-generic-with-constraints/solution.py:11]:
```suggestion
def add(a: T, b: T) -> T:
```

Otherwise the test doesn't pass

Comment (nobod4t) [reply challenges/intermediate-generic-with-constraints/solution.py:11]: done

File: challenges/intermediate-union/solution.py (hunk 2 above)
Comment (nobod4t) [reply challenges/intermediate-union/solution.py:10]: done

File: challenges/intermediate-decorator/solution.py
Comment (nobod4t) [reply challenges/intermediate-decorator/solution.py:1]: >Remove the type info from the question code for this challenge
there shouldn't be any type annotation in the function or object to be tested.

Do you mean remove type info about T: T = TypeVar("T", bound=Callable)?
I just thought this is a generic scenario. So T type is OK. Or something else you mean?

Comment (nobod4t) [reply challenges/intermediate-decorator/solution.py:1]: These are from main.  If you like I can remove them. 
Screenshot 2023-11-08 at 7 29 25 AM

Comment (frostming) [reply challenges/intermediate-decorator/solution.py:1]: yeah. please go ahead

Comment (laike9m) [reply challenges/intermediate-decorator/solution.py:1]: Thanks for fixing the question.

@frostming could you help confirm if the current solution is what you have in mind?

File: challenges/intermediate-generic-with-constraints/solution.py (hunk 1 above)
Comment (laike9m) [reply challenges/intermediate-generic-with-constraints/solution.py:9]: Could you add a comment like
`# For Python >= 3.12`

thanks

File: challenges/intermediate-generic/solution.py (hunk 3 above)
Comment (laike9m) [reply challenges/intermediate-generic/solution.py:12]: @nobod4t I updated the test case (https://github.com/laike9m/Python-Type-Challenges/commit/795aa6abcaa0cd7d39f6f79959cbe70ffe9176e0), could you pull the latest code and update the solution as suggested?

File: challenges/intermediate-decorator/question.py [hunk 5]
```diff
@@ -5,10 +5,8 @@
 """
 from typing import Callable, TypeVar
 
-T = TypeVar("T", bound=Callable)
 
-
-def decorator(func: T) -> T:
+def decorator(func: Callable) -> Callable:
```
Comment (frostming) [inline challenges/intermediate-decorator/question.py:9]:
```suggestion
def decorator(func):
```

File: challenges/intermediate-decorator/solution.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+T = TypeVar("T", bound=Callable)
+
+def decorator(func: T) -> T:
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        return func(*args, **kwargs)
+
+    return cast(T, wrapper)
```
Comment (frostming) [inline challenges/intermediate-decorator/solution.py:17]:
```suggestion
    return func
```

**Pull request merged**